app/controllers/application_controller.rb

```diff
@@ -18,8 +20,64 @@ def set_checkout_redirect
     end
   end
 
+  def enable_embedded_styles
+    session[:embedded_shopfront] = true
+    render json: {}, status: 200
+  end
+
+  def disable_embedded_styles
+    session.delete :embedded_shopfront
+    session.delete :shopfront_redirect
+    render json: {}, status: 200
+  end
+
+  protected
+
+  def after_sign_in_path_for(resource_or_scope)
+    return session[:shopfront_redirect] if session[:shopfront_redirect]
+    stored_location_for(resource_or_scope) || signed_in_root_path(resource_or_scope)
+  end
+
+  def after_sign_out_path_for(resource_or_scope)
+    session[:shopfront_redirect] ? session[:shopfront_redirect] : root_path
+  end
+
   private
 
+  def restrict_iframes
+    response.headers['X-Frame-Options'] = 'DENY'
+    response.headers['Content-Security-Policy'] = "frame-ancestors 'none'"
+  end
+
+  def enable_embedded_shopfront
+    return unless Spree::Config[:enable_embedded_shopfronts]
+
+    whitelist = Spree::Config[:embedded_shopfronts_whitelist] || "'none'"
+    response.headers.delete 'X-Frame-Options'
+    response.headers['Content-Security-Policy'] = "frame-ancestors #{whitelist}"
+
+    check_embedded_request
+    set_embedded_layout
```

Comment:
This is confusing, if there is no whitelist, but still `check_embedded_request` and `set_embedded_layout` are called.

For example - I can call `/shop?embedded_shopfront=true`, and then session variable is set unnecessary and `embedded` css class is set as as well, right?